src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/LanguageServices/CSharpTypeInferenceService.TypeInferrer.cs

```diff
@@ -1720,6 +1722,82 @@ private IEnumerable<TypeInferenceInfo> InferTypeInExpressionColon(ExpressionColo
                 }
             }
 
+            public IEnumerable<TypeInferenceInfo> InferTypeForPossibleAsyncMemberDeclarationWhenAsyncIsTreatedAsIdentifier(SyntaxToken previousToken)
+            {
+                // Edge case:
+                //
+                // class Test
+                // {
+                //   public async $$
+                // }
+                //
+                // "async" is treated as identifier name here
+
+                if (!previousToken.HasMatchingText(SyntaxKind.AsyncKeyword))
+                {
+                    return SpecializedCollections.EmptyEnumerable<TypeInferenceInfo>();
+                }
+
+                return GetPossibleTaskInference();
+            }
+
+            private IEnumerable<TypeInferenceInfo> InferTypeForPossibleAsyncMemberDeclaration(MemberDeclarationSyntax memberDeclaration, SyntaxToken previousToken)
+            {
+                // Case, when potential new member is declared before another member:
+                //
+                // class Test
+                // {
+                //    public async $$
+                //    
+                //    public void M() {}
+                // }
+                //
+                // In this case the both lines are treated as one member declaration
+
+                if (!memberDeclaration.Modifiers.Any(SyntaxKind.AsyncKeyword))
+                {
+                    return SpecializedCollections.EmptyEnumerable<TypeInferenceInfo>();
+                }
+
+                var isAfterAsync = false;
+
+                foreach (var modifier in memberDeclaration.Modifiers)
+                {
+                    if (modifier.IsKind(SyntaxKind.AsyncKeyword))
+                    {
+                        isAfterAsync = true;
+                        break;
+                    }
+
+                    if (modifier == previousToken)
+                    {
+                        break;
+                    }
+                }
+
+                if (!isAfterAsync)
+                {
+                    return SpecializedCollections.EmptyEnumerable<TypeInferenceInfo>();
+                }
+
+                return GetPossibleTaskInference();
+            }
+
+            private IEnumerable<TypeInferenceInfo> GetPossibleTaskInference()
+            {
+                var possibleTaskTypes = Compilation.GetTypesByMetadataName("System.Threading.Tasks.Task");
+                var possibleTask = possibleTaskTypes.Where(el => el.TypeParameters.Length == 0 && el.IsFromSystemRuntimeOrMscorlibAssembly()).SingleOrDefault();
+
+                if (possibleTask is null)
+                {
+                    return SpecializedCollections.EmptyEnumerable<TypeInferenceInfo>();
+                }
+
+                var inference = new TypeInferenceInfo(possibleTask);
+
+                return SpecializedCollections.SingletonEnumerable(inference);
+            }
+
```

Comment:
i would not make thsi change as part of hte type-inferrer.  THe type-inferrer is used when we're going to generate an expression and we need to figure out a reasonable type for the expression we're about to generate.  for example, if you do `if (DoesNotExist())` and we generate the method for DoesNotExist, then this would produce a return type of `bool`.  

Instead, this should be somewhere in CSharpRecommendationServiceRunner (precise location TBD).

Reply:
Type inferer is currenly the only thing, that is used to determin preselected items in completion list when generating recomendations:
https://github.com/dotnet/roslyn/blob/9c9cf66caba528d954a1c30e4f22a44e101b0e7c/src/Features/Core/Portable/Completion/Providers/AbstractRecommendationServiceBasedCompletionProvider.cs#L47-L54
So it is used not only for expression generation. IMO, it's purpose is to just offer most suitable type based on current context. `if (DoesNotExist())` example is not against this principal. Member declaration is just another case, but it keeps the purpose of type inferrer the same: we try to predict most suitable type. In case of async members it is just a `Task` as the most common one.